include/fiction/algorithms/iter/gray_code_iterator.hpp

```diff
@@ -0,0 +1,295 @@
+     * @return `true` if the current number is less than or equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator<=(const uint64_t m) const noexcept
+    {
+        return current_number <= m;
+    }
+    /**
+     * Greater-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is greater than `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator>(const uint64_t m) const noexcept
+    {
+        return current_number > m;
+    }
+    /**
+     * Greater-or-equal-than operator. Compares the current number with the given integer.
+     *
+     * @param m Integer to compare with.
+     * @return `true` if the current number is greater than or equal to `m`, `false` otherwise.
+     */
+    [[nodiscard]] bool operator>=(const uint64_t m) const noexcept
+    {
+        return current_number >= m;
+    }
+    /**
+     * Returns the start number, i.e., number for which the first Gray code is determined.
+     *
+     * @return Start number of the iteration.
+     */
+    [[nodiscard]] uint64_t get_start_number() const
+    {
+        return start_number;
+    }
+    /**
+     * Returns the current Gray code of the current iteration step.
+     *
+     * @return Current Gray code.
+     */
+    [[nodiscard]] uint64_t get_current_gray_code() const
+    {
+        return current_gray_code;
+    }
+    /**
+     * Returns the final number, i.e., final number for which the Gray code is determined.
+     *
+     * @return Final number of the iteration.
+     */
+    [[nodiscard]] uint64_t get_final_number() const
+    {
+        return final_number;
+    }
+    /**
+     * Determines if the iterator can be increased by one and does still not overcome the final number.
+     *
+     * @return `true` if the iterator can be increase by one, `false`otherwise.
+     */
+    [[nodiscard]] bool has_next() const
+    {
+        return current_number < final_number;
+    }
+
+  private:
+    /**
+     * Start number of the iteration.
+     */
+    const uint64_t start_number;
+    /**
+     * Current number (i.e., current iteration number).
+     */
+    uint64_t current_number;
+    /**
+     * Current Gray Code.
+     */
+    uint64_t current_gray_code;
+    /**
+     * Final number of the iteration.
+     */
+    const uint64_t final_number;
+    /**
+     * Converts the current decimal number into its corresponding Gray code representation.
+     *
+     * This function operates on the current decimal number and produces its Gray code
+     * representation using the bitwise XOR operation. Gray code is a binary numeral system
+     * in which two successive values differ in only one bit.
+     *
+     * The result is stored in the 'current_gray_code' variable.
+     */
+    void binary_to_gray()
+    {
+        current_gray_code = current_number ^ (current_number >> 1u);
+    }
+};
+
+}  // namespace fiction
+
```

Comment:
You can add here a specialization of this class with iterator tags in the `std` namespace to make your class compatible with STL iterator categories. Given the operators you have implemented, it looks like a bi-directional iterator to me.

You can see `aspect_ratio_iterator.hpp` for an example of how to implement this.